### PR TITLE
New GossipInfo interface method for historical gossip progress

### PR DESCRIPTION
**Pull request by maackle**
### Summary

Adds new GossipInfo method, which contains info that can be used to display current progress of historical gossip, both incoming and outgoing, per DNA.

Help me name these things better.

### TODO:
- [ ] CHANGELOG(s) updated with appropriate info
- [ ] Just before pressing the merge button, ensure new entries to CHANGELOG(s) are still under the _UNRELEASED_ heading

### CODE REVIEW

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 1]
```diff
@@ -992,31 +1015,35 @@ impl ShardedGossipLocal {
...
+                    self.queue_incoming_regions(&peer_cert, state, region_set)
+                        .await?
                 } else {
                     vec![]
                 }
             }
             ShardedGossipWire::MissingOps(MissingOps { ops, finished }) => {
+                // for op in &ops {
+                //     println!("OP RECEIVED: {} B, data: {:?}", op.0.len(), op.0);
+                // }
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:1028]:
```suggestion
```

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 2]
```diff
@@ -1139,12 +1164,36 @@ impl RoundState {
...
+        // tracing::debug!(
+        //     "FINISHED? {} {} {} {} {} {} {}",
+        //     self.num_expected_op_blooms == 0,
+        //     !self.has_pending_historical_op_data,
+        //     self.received_all_incoming_op_blooms,
+        //     self.regions_are_queued,
+        //     self.bloom_batch_cursor.is_none(),
+        //     self.ops_batch_queue.is_empty(),
+        //     ops_finished
+        // );
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:1189]:
```suggestion
```

File: crates/kitsune_p2p/kitsune_p2p/src/metrics.rs [hunk 3]
```diff
@@ -115,16 +118,29 @@ pub struct NodeInfo {
...
+pub struct PeerNodeHistory {
+    /// The most recent list of remote agents reported by this node
+    pub remote_agents: Vec<Arc<KitsuneAgent>>,
+
+    /// Detailed info about the ongoing round with this node
+    pub current_round: Option<CurrentRound>,
+
+    /// Detailed info about rounds completed with this node
+    pub completed_rounds: VecDeque<CompletedRound>,
 }
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/metrics.rs:144]: @neonphog here is the first instance of node-based as opposed to agent-based metrics

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 4]
```diff
@@ -275,6 +273,12 @@ impl ShardedGossip {
                         ShardedGossipWire::NoAgents(_) => {}
                         ShardedGossipWire::AlreadyInProgress(_) => {}
                     }
+                    i.metrics.write().update_current_round(
+                        &cert,
+                        self.gossip.gossip_type.into(),
+                        state,
+                    );
+                    // println!("throughput OUT {:?}", state.throughput);
```
Comment (maackle) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:281]:
```suggestion
```

File: crates/holochain_conductor_api/src/app_interface.rs [hunk 5]
```diff
@@ -61,9 +60,16 @@ pub enum AppRequest {
...
 
+    /// Info about gossip
+    GossipInfo(Box<GossipInfoRequestPayload>),
+
     #[deprecated = "use ZomeCall"]
     ZomeCallInvocation(Box<ZomeCall>),
 
+    /// Is currently unimplemented and will return
+    /// an [`AppResponse::Unimplemented`].
+    Crypto(Box<CryptoRequest>),
```
Comment (neonphog) [inline crates/holochain_conductor_api/src/app_interface.rs:71]: o_O

File: crates/holochain_conductor_api/src/app_interface.rs [hunk 6]
```diff
@@ -207,6 +216,37 @@ impl From<InstalledAppInfoStatus> for AppStatus {
...
+)]
+pub struct HistoricalGossipThroughput {
+    /// Total number of bytes expected to be sent for region data (historical only)
+    pub expected_op_bytes: InOut,
+    /// Total number of ops expected to be sent for region data (historical only)
+    pub expected_op_count: InOut,
+    /// Total number of ops sent
+    pub op_count: InOut,
+    /// Total number of bytes sent for op data
+    pub op_bytes: InOut,
```
Comment (neonphog) [inline crates/holochain_conductor_api/src/app_interface.rs:236]: Just sorting and spacing.

```suggestion
    /// Total number of bytes expected to be sent for region data (historical only)
    pub expected_op_bytes: InOut,

    /// Total number of ops expected to be sent for region data (historical only)
    pub expected_op_count: InOut,

    /// Total number of bytes sent for op data
    pub op_bytes: InOut,

    /// Total number of ops sent
    pub op_count: InOut,
```

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs [hunk 7]
```diff
@@ -256,10 +257,7 @@ impl ShardedGossip {
                             state.throughput.op_bloom_bytes.outgoing +=
                                 missing_hashes.size() as u32;
                         }
-                        ShardedGossipWire::OpRegions(OpRegions { region_set, .. }) => {
-                            state.throughput.region_bytes.outgoing +=
-                                region_set.regions().map(|r| r.data.size).sum::<u32>();
-                        }
+                        ShardedGossipWire::OpRegions(OpRegions { region_set: _, .. }) => {}
```
Comment (neonphog) [inline crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:260]:
```suggestion
                        ShardedGossipWire::OpRegions(OpRegions { region_set: _, .. }) => (),
```

Comment (neonphog) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:260]: Is it correct that this is no longer doing anything?

**Review by neonphog — APPROVED** [review]

File: crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs (hunk 7 above)
Comment (maackle) [reply crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip.rs:260]: Yes, at first it was doing something, and then it wasn't, so I kept the doing-nothing explicit.

File: crates/holochain_conductor_api/src/app_interface.rs (hunk 5 above)
Comment (maackle) [reply crates/holochain_conductor_api/src/app_interface.rs:71]: yeah, we could remove this...

**Comment by matthme** [comment]
@maackle I assume the order of dna's is guaranteed to be maintained between request and response? Otherwise there would seem to be no way to associate dna hash with the corresponding gossip info.

**Comment by maackle** [comment]
@matthme correct, the order is maintained

**Pull request merged**